### PR TITLE
Enforce `-` as word separator in devicetree property names

### LINKED ISSUE #53506: Require recommended word separator (`-`) in devicetree property names
## Introduction

Devicetree properties, by convention, use dash (`-`) as a word separator. This RFC proposes making this a requirement for upstream bindings.

### Problem description

There are currently no automated checks in CI that ensure that devicetree properties use dashes as separators. As a result, many properties use underscore (`_`) as a separator. This is inconsistent with the style adopted by the devicetree specification for all standard properties (with the exception of `device_type`, but this property is deprecated and has limited usage).

It would be better style to use dashes instead, and would make things consistent for users typing properties. It also would avoid situations like having a `foo-bar` and `foo_bar` property defined on the same node due to typos.

### Proposed change

1. Make all in-tree bindings and DTS files use dashes
2. Add CI checks to make sure we don't regress this

## Detailed RFC

This is a straightforward treewide change.

### Proposed change (Detailed)

1. Use grep to fix in-tree occurrences
2. Modify edtlib to expose first-class methods that list all binding objects
3. Add a new compliance check that uses these methods to complain on regressions, allowing administrator override if cases arise where we are inheriting an underscore in a property name from Linux bindings or other authoritative sources

### Dependencies

Out of tree users of existing bindings will need to migrate to the new property names.

### Concerns and Unresolved Questions

None.

## Alternatives

An approach which allowed both styles for backwards incompatibility was considered, but this is too ripe for introducing simultaneous use of `foo-bar` vs `foo_bar` in a single devicetree.

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Fixes: #53506

This tree-wide change cleans up existing property names which use `_` instead of `-`, and adds a CI check to avoid new ones from creeping in. This requires a little bit of prep work to add a helper function to edtlib to make the check easier to write.

I found a few issues in check_compliance.py while I was adding this that I took the time to fix while I was here.

### CODE REVIEW

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 1]
```diff
@@ -538,6 +539,33 @@ def _check(self):
...
 
+                self._check_compatible(node, compat)
+        del self._checked_compatibles  # We have no need for this anymore.
+
+    def _check_compatible(self, node, compat):
+        if compat in self._checked_compatibles:
+            return
+
+        # The regular expression comes from dt-schema.
+        compat_re = r'^[a-zA-Z][a-zA-Z0-9,+\-._]+$'
```
Comment (fabiobaltieri) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:550]: Should we stick to lowercase? That's what I was going on in my version of this, it matches what we have right now in the repository, would save some future incoherence with capitalization.

Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:550]: I don't think so. The regular expression we are using here is the one that dt-schema itself is using to validate compatibles. Why should we be stricter?

File: scripts/ci/check_compliance.py [hunk 2]
```diff
@@ -704,6 +708,38 @@ class KconfigBasicCheck(KconfigCheck):
...
+
+    This is NOT an exhaustive check; it should only be
+    considered a 'smoke test'.
+    """
+    name = "DevicetreeBindings"
+    doc = "See https://docs.zephyrproject.org/latest/build/dts/bindings.html for more details."
+    path_hint = ZEPHYR_BASE
+
+    def run(self):
+        bindings_dir = str(Path(ZEPHYR_BASE) / 'dts' / 'bindings')
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply scripts/ci/check_compliance.py]: > There's already a DevicetreeBindingsCheck calss here that goes over all the binding files and does few checks, can this integrate into that? 

Of course, thanks for catching that :). I botched the rebase on that part. Fixed.

> My main thought is that this filer is only checking the bindings in $ZEPHYR_BASE/dts/bindings, but the buildsystem allows extra bindings on the application directory as well and it would be cool if we could check those too.

Agree, it would be nice, but I don't think the current checks are doing that; are they?

At any rate, if we can get a list of directories to check from the calling environment, we can extend the current code to change `edtlib.bindings_from_dirs([bindings_dir])` to `edtlib.bindings_from_dirs(list_of_bindings_dirs)` very easily.

Can you suggest a way to get the list we need into that method?

[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply scripts/ci/check_compliance.py]: > > My main thought is that this filer is only checking the bindings in $ZEPHYR_BASE/dts/bindings, but the buildsystem allows extra bindings on the application directory as well and it would be cool if we could check those too.
> 
> Agree, it would be nice, but I don't think the current checks are doing that; are they?

They are right now, check is:

https://github.com/zephyrproject-rtos/zephyr/blob/f09111a2a97aa3457d2a6cf9c362c53133348d3e/scripts/ci/check_compliance.py#L245

so it looks for anything with `dts/bindings` in the path, the buildsystem code is somewhat similar:

https://github.com/zephyrproject-rtos/zephyr/blob/f09111a2a97aa3457d2a6cf9c362c53133348d3e/cmake/modules/dts.cmake#L54-L60

https://github.com/zephyrproject-rtos/zephyr/blob/f09111a2a97aa3457d2a6cf9c362c53133348d3e/cmake/modules/dts.cmake#L141

> At any rate, if we can get a list of directories to check from the calling environment, we can extend the current code to change `edtlib.bindings_from_dirs([bindings_dir])` to `edtlib.bindings_from_dirs(list_of_bindings_dirs)` very easily.
> 
> Can you suggest a way to get the list we need into that method?

I don't think we can cover all cases since one may come up with some weird combination of binding inclusions from multiple roots or custom paths, but I think for the scope of this it may be good enough to load the base bindings + a potential root of any file modified by the patch currently being tested. So we could use the `get_files(filter="d")` list to find all the potential roots (anything under a 'dts/bindings' sub path) and then do the check on "base" and any "base + external root".

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 1 above)
Comment (fabiobaltieri) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:550]: Because that is the style that we effectively adopted right now, the current set of compatibles (all 1200+ of them) are coherent in being all lowercase and the vendor names are all lowercase. It'd be nice to stick to it, avoid a code review conversation in the line of "but the standard allows it" versus "all the others are coherent".

File: scripts/ci/check_compliance.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply scripts/ci/check_compliance.py]: There's few bindings in the main repository outside of `$ZEPHYR_BASE/dts/bindings`

```
$ find -type f | grep dts/bindings | grep -v '\./dts/bindings/'
./boards/arm/arty/dts/bindings/arm,daplink-qspi-mux.yaml
./boards/arm/efr32bg_sltb010a/dts/bindings/silabs,gecko-wake-up-triggers.yaml
./boards/arm/twr_ke18f/dts/bindings/nxp,flexio.yaml
./boards/arm/actinius_icarus/dts/bindings/actinius-charger-enable.yaml
./boards/arm/actinius_icarus/dts/bindings/actinius-sim-select.yaml
./boards/arm/pinetime_devkit0/dts/bindings/misc/pine64,pinetime-key-out.yaml
./boards/arm/96b_wistrio/dts/bindings/qorvo,rfsw8001.yaml
...
```
And more in `samples` and `tests`.

The current code is considering those as well -- but I'm also interested in having this check usable in an external repository (would like to run it as part of our EC codebase presubmit), so it'd be ideal if it would just work.

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py (hunk 1 above)
Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:550]: I checked upstream Linux and there are compatibles which use uppercase letters across multiple vendors (including big names like samsung, ti, fsl, brcm, intel, national, marvell). Given that, do you still think this is a good idea?

Comment (fabiobaltieri) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:550]: Yeah I checked as well and honestly I find them to look odd, for example `"atmel,at25256B"`. I would find it nice to be strict and just stick to lowercase since we landed there naturally. Maybe seek for a third opinion? @galak?

Comment (fabiobaltieri) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py:550]: Another thought: if we try and be strict now and then it turns out we change our mind, it's easy to change the check to be less strict. Not so much the other way around.

Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/edtlib.py]: OK, I guess we can start there for now, but I'm going to resolve this comment and move the compatibles check to another PR now that we've got the tree-wide `s/_/-/g` change approved.

File: scripts/ci/check_compliance.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply scripts/ci/check_compliance.py]: > The current code is considering those as well

I restored this functionality; thanks.

> I'm also interested in having this check usable in an external repository

Fair enough, but that's out of scope for what I'm trying to do since the current code isn't doing that.

**Comment by mbolivar-nordic** [comment]
@de-nordic I need https://github.com/zephyrproject-rtos/fatfs/pull/16 for this PR

File: boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts [hunk 3]
```diff
@@ -69,8 +69,7 @@
 };
 
 &i2c_smb_0 {
-	status = "okay";
```
Comment (henrikbrixandersen) [inline boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts:72]: Was this removal intentional?

File: boards/arm/mec15xxevb_assy6853/mec15xxevb_assy6853.dts [hunk 4]
```diff
@@ -131,8 +131,7 @@
 };
 
 &i2c_smb_2 {
-	status = "okay";
```
Comment (henrikbrixandersen) [inline boards/arm/mec15xxevb_assy6853/mec15xxevb_assy6853.dts:134]: Same here?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, couple of nits:

File: scripts/ci/check_compliance.py [hunk 5]
```diff
@@ -34,6 +35,21 @@
                str((HERE / '..' / '..').resolve()))
 os.environ['ZEPHYR_BASE'] = ZEPHYR_BASE
 
+@contextlib.contextmanager
+def run_in_directory(dir):
+    # Helper to run a code block from a given directory.
+    # Use it like this:
+    #
+    #     with run_in_directory('some/directory'):
+    #         do_something()
```
Comment (fabiobaltieri) [inline scripts/ci/check_compliance.py:48]: nit: can you make this pydoc comment instead?

**Review by fabiobaltieri — COMMENTED** [review]
Cool, thanks Martí. Thinking we could add some more strict test as a start but this looks like a solid foundation to build up upon.

**Comment by de-nordic** [comment]
@henrikbrixandersen Sorry for calling you to the re-review, somehow I managed to press the button in wrong review.

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by MaureenHelm — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
Previous force push is a rebase and removal of the west.yml patch

**Comment by mbolivar-nordic** [comment]
@henrikbrixandersen thank you for catching the `status = "okay";` removals; those were artifacts of a botched rebase. Fixed.

**Comment by henrikbrixandersen** [comment]
@mbolivar-nordic At least the compliance check works 😂

**Review by henrikbrixandersen — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

File: scripts/ci/check_compliance.py [hunk 6]
```diff
@@ -25,8 +25,15 @@
 sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
 from get_maintainer import Maintainers, MaintainersError
 
+HERE = Path(__file__).parent
+
 logger = None
 
+# This ends up using the current repository if ZEPHYR_BASE is unset.
```
Comment (carlescufi) [inline scripts/ci/check_compliance.py:34]: But we already have a default for `ZEPHYR_BASE`:

https://github.com/zephyrproject-rtos/zephyr/blob/ee9903005a6cebdce4c0cc5b8bf8144c9ec8ab7c/scripts/ci/check_compliance.py#L1110-L1118

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
You wrote:

> 1. compliance.yml happens to run the script from ZEPHYR_BASE, which
   is what the tests mostly seem to want anyway

But that's only true upstream. Downstream this is run from the root of the repo that the checks need to be run on.

File: scripts/ci/check_compliance.py (hunk 6 above)
Comment (mbolivar-nordic) [reply scripts/ci/check_compliance.py:34]: > But we already have a default for ZEPHYR_BASE:

Oh, I didn't see that this function was changing global variables. That's fun.

I still need this at module level so that `check_compliance: re-work DevicetreeBindingsCheck` later on in the PR works, so I'll just refactor this patch to remove the lines you mention as well.

**Comment by mbolivar-nordic** [comment]
> But that's only true upstream. Downstream this is run from the root of the repo that the checks need to be run on.

You're saying that we are running the zephyr/.github/workflows/compliance.yml file from non-zephyr repositories in downstreams?

If so, I'm having trouble understanding if you're asking for a change, though. The documented `path_hint` behavior is

```
    path_hint:
      The path the test runs itself in.  This is just informative and used in
      the message that gets printed when running the test.
```

and this PR is making the implementation match this documentation, so I don't see any changes that need doing, except maybe removing the "just informative" part. Am I missing something?

**Comment by mbolivar-nordic** [comment]
@carlescufi ping, can you please explain what the changes you are requesting are? See my previous comment.

**Comment by fabiobaltieri** [comment]
@carlescufi ping? It'd be cool to move forward with this.

**Review by cfriedt — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
This needs rebasing.

**Comment by cfriedt** [comment]
@mbolivar-nordic - please rebase and resolve conflicts. I'll approve again

**Comment by henrikbrixandersen** [comment]
Removing the `stale` label as I hope this will be picked up again.

**Comment by carlescufi** [comment]
@mbolivar-nordic apologies for the huge delay. Your patches make sense, in general, I will add a comment in the code itself.

File: scripts/ci/check_compliance.py (hunk 6 above)
Comment (carlescufi) [reply scripts/ci/check_compliance.py:34]: But you did not remove those lines, did you?

**Comment by carlescufi** [comment]
> You're saying that we are running the zephyr/.github/workflows/compliance.yml file from non-zephyr repositories in downstreams?

No, not the same exact workflow, but a very similar one. See here:
https://github.com/nrfconnect/sdk-nrf/blob/7b40e24932178c3686fcd3dee71796b7b7962dc7/.github/workflows/compliance.yml#L78

**Comment by cfriedt** [comment]
This would be a good thing to redo before lts3

**Comment by henrikbrixandersen** [comment]
> This would be a good thing to redo before lts3

Agreed. Would love to see this happen.
